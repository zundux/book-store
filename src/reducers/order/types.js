export const SAVE_ORDER_REQUEST = "order/SAVE_ORDER_REQUEST";
export const SAVE_ORDER_SUCCESS = "order/SAVE_ORDER_SUCCESS";
export const SAVE_ORDER_FAIL = "order/SAVE_ORDER_FAIL";