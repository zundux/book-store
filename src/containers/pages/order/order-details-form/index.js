import OrderDetailsForm from "./order-details-form";

export default OrderDetailsForm;