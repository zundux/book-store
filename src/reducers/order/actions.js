import fetch from "cross-fetch";
import * as R from "ramda";
import { push } from "connected-react-router";
import { SAVE_ORDER_FAIL, SAVE_ORDER_REQUEST, SAVE_ORDER_SUCCESS } from "./types";
import { API_BASE_URL } from "../../api";
import { cleanCart } from "../cart";

const orderResourceUrl = `${API_BASE_URL}/orders`;

export const saveOrderDetails = (orderDetails, products) => ({
  type: SAVE_ORDER_REQUEST,
  orderDetails,
  products
});

export const receiveOrderSuccess = ({ id }) => ({
  type: SAVE_ORDER_SUCCESS,
  id
});

export const receiveOrderFail = ({ message }) => ({
  type: SAVE_ORDER_FAIL,
  errorMessage: `Could not save order. ${message}`
});


export const saveOrder = (orderDetails, products) => {
  return (dispatch, getState )=> {
    dispatch(saveOrderDetails(orderDetails, products));
    console.log(getState());

    return sendSaveOrderRequest(orderDetails, products, dispatch);
  };
};

export const retrySaveOrder = () => {
  return (dispatch, getState) => {
    const { order: { orderDetails, products } } = getState();

    return sendSaveOrderRequest(orderDetails, products, dispatch);
  };
};


const orderSavedSuccessfully = dispatch => json => {
  const a = dispatch(receiveOrderSuccess(json));
  dispatch(push("/order/created"));
  cleanCart()(dispatch);

  return a;
};

const orderSaveFailure = dispatch => error => dispatch(receiveOrderFail(error));

function sendSaveOrderRequest(orderDetails, products, dispatch) {
  const productIds = R.pluck("id", products);

  return fetch(orderResourceUrl, {
    headers: {
      "content-type": "application/json"
    },
    method: "POST",
    mode: "CORS",
    credentials: "include",
    body: JSON.stringify({ orderDetails, productIds })
  })
    .then(response => response.json())
    .then(orderSavedSuccessfully(dispatch))
    .catch(orderSaveFailure(dispatch));
}