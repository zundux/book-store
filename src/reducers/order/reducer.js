import { combineReducers } from "redux";
import { SAVE_ORDER_FAIL, SAVE_ORDER_REQUEST, SAVE_ORDER_SUCCESS } from "./types";

export const initialStateOrder = {
  id: null,
  products: [],
  orderDetails: null,

  isSending: false,
  errorMessage: null
};

const id = (state = null, action) => {
  switch (action.type) {
    case SAVE_ORDER_REQUEST:
      return null;

    case SAVE_ORDER_SUCCESS:
      return action.id;

    case SAVE_ORDER_FAIL:
    default:
      return state;
  }
};

const errorMessage = (state = null, action) => {
  switch (action.type) {
    case SAVE_ORDER_REQUEST:
    case SAVE_ORDER_SUCCESS:
      return null;

    case SAVE_ORDER_FAIL:
      return action.errorMessage;

    default:
      return state;
  }
};

const isSending = (state = false, action) => {
  switch (action.type) {
    case SAVE_ORDER_REQUEST:
      return true;

    case SAVE_ORDER_SUCCESS:
    case SAVE_ORDER_FAIL:
      return false;

    default:
      return state;
  }
};

const products = (state = [], action) => {
  switch (action.type) {
    case SAVE_ORDER_REQUEST:
      return action.products;

    case SAVE_ORDER_SUCCESS:
      return [];

    case SAVE_ORDER_FAIL:
    default:
      return state;
  }
};

const orderDetails = (state = null, action) => {
  switch (action.type) {
    case SAVE_ORDER_REQUEST:
      return action.orderDetails;

    case SAVE_ORDER_SUCCESS:
      return null;

    case SAVE_ORDER_FAIL:
    default:
      return state;
  }
};

export default combineReducers({
  id,
  products,
  orderDetails,
  isSending,
  errorMessage
});