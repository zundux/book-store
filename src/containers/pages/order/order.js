import React from "react";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import OrderDetailsForm from "./order-details-form";
import { retrySaveOrder, saveOrder } from "../../../reducers/order";
import { initialValuesOfOrderDetailsForm, validateOrderDetails } from "../../../reducers/order-details";
import ErrorLabel from "../../../components/error-label";

export const Order = ({ errorMessage, products, onSaveOrder, onRetrySaveOrder }) => {
  const onSubmit = formValues => onSaveOrder(formValues, products);

  return (
    <div>
      <h1>Order details</h1>
      <OrderDetailsForm
        onSubmit={onSubmit}
        initialValues={initialValuesOfOrderDetailsForm}
        validate={validateOrderDetails}
      />
      {errorMessage && (
        <ErrorLabel
          message={errorMessage}
          onRetry={onRetrySaveOrder}
        />
      )}
    </div>
  );
};

const mapStateToProps = ({ cart, order }) => ({
  products: cart.products,
  errorMessage: order.errorMessage
});

const mapDispatchToProps = dispatch => bindActionCreators({
  onSaveOrder: saveOrder,
  onRetrySaveOrder: retrySaveOrder
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Order);