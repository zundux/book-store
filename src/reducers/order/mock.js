import { initialStateOrder } from "./reducer";

export const saveOrderData = {
  products: [
    { id: 1, title: "book 1", amount: 1 },
    { id: 2, title: "book 2", amount: 2 }
  ],
  orderDetails: {
    contactName: "Joe Doe",
    phoneNumber: "+1 212-340-0863",
    email: "example@example.com",
    paymentAddress: "476 5th Ave, New York, NY 10018, USA"
  }
};

export const errorMessage = "Something went wrong";

export default {
  initialState: initialStateOrder,
  saveOrderData,
  errorMessage
};