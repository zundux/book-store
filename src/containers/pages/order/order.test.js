import React from "react";
import { Order } from "./order";

describe("Order page", () => {

  it("Order page should not show error message by default", () => {
    const products = [];
    const errorMessage = null;
    const onSaveOrder = () => {
    };
    const onRetrySaveOrder = () => {
    };

    const component = shallow(
      <Order {...{
        products,
        errorMessage,
        onSaveOrder,
        onRetrySaveOrder
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
  });

  it("Order page should show error message if something went wrong", () => {
    const errorMessage = "some error message";
    const products = [];
    const onSaveOrder = () => {
    };
    const onRetrySaveOrder = () => {
    };

    const component = shallow(
      <Order {...{
        products,
        errorMessage,
        onSaveOrder,
        onRetrySaveOrder
      }}/>
    );
    matchSnapshot(component);
    // todo check page header
    // todo check page title
  });
});