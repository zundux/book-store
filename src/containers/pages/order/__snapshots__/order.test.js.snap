// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order page Order page should not show error message by default 1`] = `
<div>
  <h1>
    Order details
  </h1>
  <Connect(ReduxForm)
    initialValues={
      Object {
        "contact": Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "phone": "",
        },
        "payment": Object {
          "address": "",
          "addressOptional": "",
          "city": "",
          "country": "",
          "postCode": "",
          "region": "",
        },
        "shipping": Object {
          "address": "",
          "addressOptional": "",
          "city": "",
          "country": "",
          "postCode": "",
          "region": "",
        },
        "useShippingInfoAsPaymentInfo": true,
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  />
</div>
`;

exports[`Order page Order page should show error message if something went wrong 1`] = `
<div>
  <h1>
    Order details
  </h1>
  <Connect(ReduxForm)
    initialValues={
      Object {
        "contact": Object {
          "email": "",
          "firstName": "",
          "lastName": "",
          "phone": "",
        },
        "payment": Object {
          "address": "",
          "addressOptional": "",
          "city": "",
          "country": "",
          "postCode": "",
          "region": "",
        },
        "shipping": Object {
          "address": "",
          "addressOptional": "",
          "city": "",
          "country": "",
          "postCode": "",
          "region": "",
        },
        "useShippingInfoAsPaymentInfo": true,
      }
    }
    onSubmit={[Function]}
    validate={[Function]}
  />
  <ErrorLabel
    message="some error message"
    onRetry={[Function]}
  />
</div>
`;
